src/backend/distributed/commands/utility_hook.c

```diff
@@ -644,6 +666,179 @@ ProcessUtilityInternal(PlannedStmt *pstmt,
 		 */
 		CitusHasBeenLoaded();
 	}
+
+	if (list_length(ddlJobs) != 0)
+	{
+		return true;
+	}
+
+	if (IsDropSchemaOrTable(parsetree))
+	{
+		/*
+		 * TODO: We could specifically check if it is a:
+		 * - DROP TABLE statement that drops a reference or a citus local table
+		 *   that has a foreign key relationship with a postgres table, or a
+		 * - DROP SCHEMA statement that again drops such a citus table
+		 * but this is more easy and the other approach is much more prune to
+		 * deadlocks.
+		 */
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * IsDropSchemaOrTable returns true if parsetree represents a DROP SCHEMA
+ * or a DROP TABLE statement.
+ */
+static bool
+IsDropSchemaOrTable(Node *parsetree)
+{
+	if (!IsA(parsetree, DropStmt))
+	{
+		return false;
+	}
+
+	DropStmt *dropStatement = (DropStmt *) parsetree;
+	ObjectType dropType = dropStatement->removeType;
+	return dropType == OBJECT_SCHEMA || dropType == OBJECT_TABLE;
+}
+
+
+/*
+ * UndistributeCitusLocalTablesIfNeeded undistributes citus local tables that
+ * are not connected to any reference tables via their individual foreign key
+ * subgraphs.
+ */
+static void
+UndistributeCitusLocalTablesIfNeeded(bool executedDDLJob)
+{
+	if (!ShouldUndistributeCitusLocalTables(executedDDLJob))
+	{
+		return;
+	}
+
+	List *citusLocalTableIdList = CitusTableTypeIdList(CITUS_LOCAL_TABLE);
+	citusLocalTableIdList = SortList(citusLocalTableIdList, CompareOids);
+
+	/* acquire ExclusiveLock as UndistributeTable does */
+	LockRelationsWithLockMode(citusLocalTableIdList, ExclusiveLock);
+
+	Oid citusLocalTableId = InvalidOid;
+	foreach_oid(citusLocalTableId, citusLocalTableIdList)
+	{
+		HeapTuple heapTuple =
+			SearchSysCache1(RELOID, ObjectIdGetDatum(citusLocalTableId));
+		if (!HeapTupleIsValid(heapTuple))
+		{
+			/*
+			 * UndistributeTable drops relation, skip if already undistributed
+			 * via cascade.
+			 */
+			continue;
+		}
+		ReleaseSysCache(heapTuple);
+
+		if (ConnectedToReferenceTableViaFKey(citusLocalTableId))
+		{
+			/* still connected to a reference table, skip it */
+			continue;
+		}
+
+		/*
+		 * Citus local table is not connected to any reference table, then
+		 * undistribute it via cascade.
+		 */
+		TableConversionParameters params = {
+			.relationId = citusLocalTableId,
+			.cascadeViaForeignKeys = true
```

Comment:
This logic seems slightly off. We already get all Citus local tables, and still use `cascadeViaForeignKeys = true`. Could this be the reason that you see `relation not exists ` errors?


(Our initial idea was that we would be able to find the exact relation which is affected by the drop, that's why we ended up `cascadeViaForeignKeys `. With the current approach, we shouldn't need this?)

Reply:
we might still have citus local tables connected to each other but not to a reference table, so I guess we need to `cascadeViaForeignKeys `

Reply:
Well, those should be qualified to be dropped individually as `ConnectedToReferenceTableViaFKey` will return false for all of them?

Reply:
yes but we will still need to frop foreign key constraints on each of such citus local tables, then it would nearly be copy/paste of cascade via foreign keys logic

Reply:
> Yes but we will still need to drop foreign key constraints on each of such citus local tables,

Ok, this should be in the comment. Plus, do we have a test covers this scenario? I think we should

Reply:
yeah I will add such tests